### PR TITLE
Update verify tooling and fixup lint warnings

### PR DESCRIPTION
**Pull request by justaugustus**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

- dependencies.yaml: Update verify tooling to latest versions
  - Update k/repo-infra to v0.2.2
  - Update zeitgeist to v0.3.0
  - Update golangci-lint to v1.41.1
- .golangci.yml: Replace `golint` (archived) with `revive`
- golangci-lint: Fixup warnings (first pass)
- stream/http.go: nolint:bodyclose for Produce()
- inventory: Minor simplification of ParseContainerParts logic
- test-e2e: Fixup lint warnings
- audit: nolint:errcheck for write errors
- checks: Fixup lint warnings
- inventory: Fixup lint warnings

#### Special notes for your reviewer:

/hold as golangci-lint is throwing a few failures locally

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

**Comment by justaugustus** [comment]
> /hold as golangci-lint is throwing a few failures locally

Odd that [these errors](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_k8s-container-image-promoter/394/pull-cip-verify/1424402321466462208) got swallowed previously:

```console
 level=warning msg="[runner] The linter 'golint' is deprecated (since v1.41.0) due to: The repository of the linter has been archived by the owner.  Replaced by revive."
legacy/dockerregistry/checks.go:330:12: Error return value is not checked (errcheck)
			edge := req.RequestParams.(PromotionEdge)
			        ^
legacy/dockerregistry/inventory.go:1366:16: Error return value is not checked (errcheck)
					parentRC, _ := req.RequestParams.(RegistryContext)
					          ^
legacy/dockerregistry/inventory.go:1399:17: Error return value of `sc.ExecRequests` is not checked (errcheck)
	sc.ExecRequests(populateRequests, processRequest)
	               ^
legacy/dockerregistry/inventory.go:1481:12: Error return value is not checked (errcheck)
			gmlc := req.RequestParams.(GCRManifestListContext)
			        ^
legacy/dockerregistry/inventory.go:1494:17: Error return value of `sc.ExecRequests` is not checked (errcheck)
	sc.ExecRequests(populateRequests, processRequest)
	               ^
legacy/dockerregistry/inventory_test.go:1792:4: Error return value of `` is not checked (errcheck)
			_ = test.input[i].Finalize()
			^
legacy/audit/auditor.go:199:6: Error return value of `w.Write` is not checked (errcheck)
		_, _ = w.Write([]byte(msg))
		   ^
legacy/audit/auditor.go:209:6: Error return value of `w.Write` is not checked (errcheck)
		_, _ = w.Write([]byte(msg))
		   ^
legacy/audit/auditor.go:248:7: Error return value of `w.Write` is not checked (errcheck)
			_, _ = w.Write([]byte(msg))
			   ^
legacy/dockerregistry/checks.go:264:2: hugeParam: syncContext is heavy (136 bytes); consider passing it by pointer (gocritic)
	syncContext SyncContext,
	^
legacy/dockerregistry/inventory.go:1183:3: emptyFallthrough: remove empty case containing only fallthrough to default case (gocritic)
		fallthrough
		^
legacy/dockerregistry/inventory.go:1181:3: emptyFallthrough: remove empty case containing only fallthrough to default case (gocritic)
		fallthrough
		^
legacy/dockerregistry/inventory.go:1174:3: emptyFallthrough: replace empty case containing only fallthrough with expression list (gocritic)
		fallthrough
		^
legacy/dockerregistry/inventory.go:1172:3: emptyFallthrough: replace empty case containing only fallthrough with expression list (gocritic)
		fallthrough
		^
legacy/dockerregistry/inventory.go:1170:3: emptyFallthrough: replace empty case containing only fallthrough with expression list (gocritic)
		fallthrough
		^
legacy/dockerregistry/inventory.go:1164:6: emptyStringTest: replace `len(part) == 0` with `part == ""` (gocritic)
		if len(part) == 0 {
		   ^
legacy/dockerregistry/inventory.go:1431:6: nestingReduce: invert if cond, replace body with `continue`, move old body after the statement (gocritic)
					if sc.DigestMediaType[digest] == ggcrV1Types.DockerManifestList {
					^
legacy/dockerregistry/inventory.go:2394:6: nestingReduce: invert if cond, replace body with `continue`, move old body after the statement (gocritic)
					if len(tagArray) == 0 {
					^
legacy/dockerregistry/inventory.go:623:6: rangeValCopy: each iteration copies 192 bytes (consider pointers or indexing) (gocritic)
					for _, edge := range edgeList {
					^
legacy/dockerregistry/inventory.go:634:5: rangeValCopy: each iteration copies 192 bytes (consider pointers or indexing) (gocritic)
				for _, edge := range edgeList {
				^
legacy/dockerregistry/inventory.go:1485:21: typeUnparen: could simplify (Digest) to Digest (gocritic)
				sc.ParentDigest[(Digest)((gManifest.Digest.Algorithm)+":"+(gManifest.Digest.Hex))] = gmlc.Digest
				                ^
legacy/dockerregistry/inventory.go:2585:16: typeUnparen: could simplify (func(ggcrV1Types.MediaType) func(ggcrV1Types.MediaType) bool) to func(ggcrV1Types.MediaType) func(ggcrV1Types.MediaType) bool (gocritic)
	var isEqualTo (func(ggcrV1Types.MediaType) func(ggcrV1Types.MediaType) bool) = func(want ggcrV1Types.MediaType) func(ggcrV1Types.MediaType) bool {
	              ^
legacy/dockerregistry/inventory.go:2591:19: typeUnparen: could simplify (func(ggcrV1Types.MediaType) func(ggcrV1Types.MediaType) bool) to func(ggcrV1Types.MediaType) func(ggcrV1Types.MediaType) bool (gocritic)
	var isNotEqualTo (func(ggcrV1Types.MediaType) func(ggcrV1Types.MediaType) bool) = func(want ggcrV1Types.MediaType) func(ggcrV1Types.MediaType) bool {
	                 ^
legacy/dockerregistry/inventory.go:1156:1: unnamedResult: consider giving a name to these results (gocritic)
func ParseContainerParts(s string) (string, string, error) {
^
test-e2e/cip/e2e.go:132:33: hugeParam: t is heavy (88 bytes); consider passing it by pointer (gocritic)
func testSetup(repoRoot string, t E2ETest) error {
                                ^
test-e2e/cip-auditor/cip-auditor-e2e.go:258:2: hugeParam: t is heavy (152 bytes); consider passing it by pointer (gocritic)
	t E2ETest,
	^
test-e2e/cip-auditor/cip-auditor-e2e.go:159:2: rangeValCopy: each iteration copies 152 bytes (consider pointers or indexing) (gocritic)
	for _, t := range ts {
	^
legacy/dockerregistry/checks.go:335: File is not `gofumpt`-ed (gofumpt)
					Error:   err})
legacy/dockerregistry/inventory.go:1444: File is not `gofumpt`-ed (gofumpt)
							Digest:          digest}
legacy/dockerregistry/inventory.go:2205: File is not `gofumpt`-ed (gofumpt)
	var populateRequests = MKPopulateRequestsForPromotionEdges(
test-e2e/cip-auditor/cip-auditor-e2e.go:257:12: `testSetup` - `pushRepo` is unused (unparam)
	repoRoot, pushRepo, projectNumber, projectID string,
	          ^
legacy/dockerregistry/inventory.go:247: unnecessary leading newline (whitespace)
		err error) error {
legacy/dockerregistry/inventory.go:577: unnecessary leading newline (whitespace)
	edges map[PromotionEdge]interface{}) (map[PromotionEdge]interface{}, error) {
legacy/dockerregistry/inventory.go:1268: unnecessary leading newline (whitespace)
		wg *sync.WaitGroup) {
legacy/stream/http.go:48:24: response body must be closed (bodyclose)
	h.Res, err = client.Do(h.Req)
	                      ^
legacy/dockerregistry/inventory.go:824:24: S1039: unnecessary use of fmt.Sprintf (gosimple)
			errs = append(errs, fmt.Sprintf("cannot have more than 1 source registry"))
			                    ^
legacy/dockerregistry/inventory.go:829:24: S1039: unnecessary use of fmt.Sprintf (gosimple)
			errs = append(errs, fmt.Sprintf("source registry must be set"))
			                    ^
legacy/dockerregistry/inventory.go:835:23: S1039: unnecessary use of fmt.Sprintf (gosimple)
		errs = append(errs, fmt.Sprintf("'registries' field cannot be empty"))
		                    ^
legacy/dockerregistry/inventory.go:1187:1: label error has same name as predeclared identifier (predeclared)
error:
^
legacy/dockerregistry/inventory.go:844:21: SA4010: this result of append is never used, except maybe in other appends (staticcheck)
		knownRegistries = append(knownRegistries, registry.Name)
		                  ^
legacy/dockerregistry/inventory.go:1707:11: ST1005: error strings should not be capitalized (stylecheck)
				err = fmt.Errorf("Encountered an error while executing requests")
				      ^ 
```

**Comment by justaugustus** [comment]
Next batch of [errors](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_k8s-container-image-promoter/394/pull-cip-verify/1424429388220338176), as of https://github.com/kubernetes-sigs/k8s-container-image-promoter/commit/0c0c0af270bb4cba8709b89545a9075af5a97f55:

<details>

```console
 legacy/dockerregistry/checks.go:331:12: Error return value is not checked (errcheck)
			edge := req.RequestParams.(PromotionEdge)
			        ^
legacy/dockerregistry/inventory.go:1376:16: Error return value is not checked (errcheck)
					parentRC, _ := req.RequestParams.(RegistryContext)
					          ^
legacy/dockerregistry/inventory.go:1411:17: Error return value of `sc.ExecRequests` is not checked (errcheck)
	sc.ExecRequests(populateRequests, processRequest)
	               ^
legacy/dockerregistry/inventory.go:1498:12: Error return value is not checked (errcheck)
			gmlc := req.RequestParams.(GCRManifestListContext)
			        ^
legacy/dockerregistry/inventory.go:1511:17: Error return value of `sc.ExecRequests` is not checked (errcheck)
	sc.ExecRequests(populateRequests, processRequest)
	               ^
legacy/dockerregistry/inventory_test.go:1792:4: Error return value of `` is not checked (errcheck)
			_ = test.input[i].Finalize()
			^
legacy/audit/auditor.go:199:6: Error return value of `w.Write` is not checked (errcheck)
		_, _ = w.Write([]byte(msg))
		   ^
legacy/audit/auditor.go:209:6: Error return value of `w.Write` is not checked (errcheck)
		_, _ = w.Write([]byte(msg))
		   ^
legacy/audit/auditor.go:248:7: Error return value of `w.Write` is not checked (errcheck)
			_, _ = w.Write([]byte(msg))
			   ^
legacy/dockerregistry/checks.go:264:2: hugeParam: syncContext is heavy (136 bytes); consider passing it by pointer (gocritic)
	syncContext SyncContext,
	^
legacy/dockerregistry/inventory.go:1191:3: emptyFallthrough: remove empty case containing only fallthrough to default case (gocritic)
		fallthrough
		^
legacy/dockerregistry/inventory.go:1189:3: emptyFallthrough: remove empty case containing only fallthrough to default case (gocritic)
		fallthrough
		^
legacy/dockerregistry/inventory.go:1182:3: emptyFallthrough: replace empty case containing only fallthrough with expression list (gocritic)
		fallthrough
		^
legacy/dockerregistry/inventory.go:1180:3: emptyFallthrough: replace empty case containing only fallthrough with expression list (gocritic)
		fallthrough
		^
legacy/dockerregistry/inventory.go:1178:3: emptyFallthrough: replace empty case containing only fallthrough with expression list (gocritic)
		fallthrough
		^
legacy/dockerregistry/inventory.go:623:6: rangeValCopy: each iteration copies 192 bytes (consider pointers or indexing) (gocritic)
					for _, edge := range edgeList {
					^
legacy/dockerregistry/inventory.go:634:5: rangeValCopy: each iteration copies 192 bytes (consider pointers or indexing) (gocritic)
				for _, edge := range edgeList {
				^
legacy/dockerregistry/inventory.go:1162:1: unnamedResult: consider giving a name to these results (gocritic)
func ParseContainerParts(s string) (string, string, error) {
^
test-e2e/cip/e2e.go:132:33: hugeParam: t is heavy (88 bytes); consider passing it by pointer (gocritic)
func testSetup(repoRoot string, t E2ETest) error {
                                ^
test-e2e/cip-auditor/cip-auditor-e2e.go:255:44: hugeParam: t is heavy (152 bytes); consider passing it by pointer (gocritic)
func testSetup(repoRoot, projectID string, t E2ETest) error {
                                           ^
test-e2e/cip-auditor/cip-auditor-e2e.go:158:2: rangeValCopy: each iteration copies 152 bytes (consider pointers or indexing) (gocritic)
	for _, t := range ts {
	^
legacy/dockerregistry/inventory.go:2224: File is not `gofumpt`-ed (gofumpt)
	var populateRequests = MKPopulateRequestsForPromotionEdges(
legacy/dockerregistry/inventory.go:2528: File is not `gofumpt`-ed (gofumpt)
	var deleteRequestsPopulator = func(
legacy/dockerregistry/inventory.go:2616: File is not `gofumpt`-ed (gofumpt)
	var isEqualTo = func(want ggcrV1Types.MediaType) func(ggcrV1Types.MediaType) bool {
legacy/stream/http.go:48:24: response body must be closed (bodyclose)
	h.Res, err = client.Do(h.Req)
	                      ^
legacy/dockerregistry/inventory.go:846:21: SA4010: this result of append is never used, except maybe in other appends (staticcheck)
		knownRegistries = append(knownRegistries, registry.Name)
		                  ^
```

</details>

**Comment by justaugustus** [comment]
Interesting to see a flake in `reqcounter` [here](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_k8s-container-image-promoter/394/pull-cip-unit-tests/1424447338000158720):

```console
 === RUN   TestCycle
    reqcounter_test.go:250: 
        	Error Trace:	reqcounter_test.go:250
        	Error:      	Not equal: 
        	            	expected: []string{"From 2006-01-02 15:04:05 to 2006-01-03 15:04:05 [1440 min] there have been 2 requests to GCR.", "From 2006-01-03 15:04:05 to 2006-01-04 15:04:05 [1440 min] there have been 622 requests to GCR.", "From 2006-01-04 15:04:05 to 2006-01-05 15:04:05 [1440 min] there have been 5 requests to GCR.", "From 2006-01-05 15:04:05 to 2006-01-06 15:04:05 [1440 min] there have been 8 requests to GCR."}
        	            	actual  : []string{"From 2006-01-02 15:04:05 to 2006-01-03 15:04:05 [1440 min] there have been 2 requests to GCR.", "From 2006-01-03 15:04:05 to 2006-01-04 15:14:05 [1440 min] there have been 622 requests to GCR.", "From 2006-01-04 15:14:05 to 2006-01-05 15:14:05 [1440 min] there have been 5 requests to GCR.", "From 2006-01-05 15:14:05 to 2006-01-06 15:14:05 [1440 min] there have been 8 requests to GCR."}
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -2,5 +2,5 @@
        	            	  (string) (len=93) "From 2006-01-02 15:04:05 to 2006-01-03 15:04:05 [1440 min] there have been 2 requests to GCR.",
        	            	- (string) (len=95) "From 2006-01-03 15:04:05 to 2006-01-04 15:04:05 [1440 min] there have been 622 requests to GCR.",
        	            	- (string) (len=93) "From 2006-01-04 15:04:05 to 2006-01-05 15:04:05 [1440 min] there have been 5 requests to GCR.",
        	            	- (string) (len=93) "From 2006-01-05 15:04:05 to 2006-01-06 15:04:05 [1440 min] there have been 8 requests to GCR."
        	            	+ (string) (len=95) "From 2006-01-03 15:04:05 to 2006-01-04 15:14:05 [1440 min] there have been 622 requests to GCR.",
        	            	+ (string) (len=93) "From 2006-01-04 15:14:05 to 2006-01-05 15:14:05 [1440 min] there have been 5 requests to GCR.",
        	            	+ (string) (len=93) "From 2006-01-05 15:14:05 to 2006-01-06 15:14:05 [1440 min] there have been 8 requests to GCR."
        	            	 }
        	Test:       	TestCycle
        	Messages:   	The request counter produced malformed logs.
--- FAIL: TestCycle (0.00s)
FAIL
coverage: 100.0% of statements
FAIL	sigs.k8s.io/k8s-container-image-promoter/legacy/reqcounter	0.015s 
```

**Comment by justaugustus** [comment]
(Moving the initial commits out to https://github.com/kubernetes-sigs/k8s-container-image-promoter/pull/395.)

**Comment by justaugustus** [comment]
After breaking up some of the commits, I've found that the changes in https://github.com/kubernetes-sigs/k8s-container-image-promoter/pull/394/commits/16150b789862b2ea349257e3a6ad9834ed87b196 lead to the [unit](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_k8s-container-image-promoter/394/pull-cip-unit-tests/1424641363101618176) and [e2e](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_k8s-container-image-promoter/394/pull-cip-e2e/1424641362984177664) test failures.

Going to drop this behind a `nolint` as well.

**Comment by justaugustus** [comment]
/hold cancel
/assign @saschagrunert @cpanato @puerco 
cc: @kubernetes-sigs/release-engineering

**Review by saschagrunert — APPROVED** [review]

File: legacy/dockerregistry/inventory.go [hunk 1]
```diff
@@ -1152,39 +1165,40 @@ func (sc *SyncContext) IgnoreFromPromotion(regName RegistryName) {
...
-			goto error
+			goto InvalidString
 		}
 		return strings.Join(parts[0:2], "/"), strings.Join(parts[2:], "/"), nil
-	case "k8s.gcr.io":
-		fallthrough
-	case "staging-k8s.gcr.io":
-		fallthrough
 	default:
+		if parts[0] != "k8s.gcr.io" && parts[0] != "staging-k8s.gcr.io" {
```
Comment (listx) [inline legacy/dockerregistry/inventory.go:1194]: Sorry I wasn't able to review this code.

I'd rather have the branchless case statements with fallthrough than have the 2 `!=` conditionals with `&&`. Can we avoid this and still call InvalidString if it's some unexpected string?

**Comment by listx** [comment]
> Interesting to see a flake in `reqcounter` [here](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_k8s-container-image-promoter/394/pull-cip-unit-tests/1424447338000158720):
> 
> ```
>  === RUN   TestCycle
>     reqcounter_test.go:250: 
>         	Error Trace:	reqcounter_test.go:250
>         	Error:      	Not equal: 
>         	            	expected: []string{"From 2006-01-02 15:04:05 to 2006-01-03 15:04:05 [1440 min] there have been 2 requests to GCR.", "From 2006-01-03 15:04:05 to 2006-01-04 15:04:05 [1440 min] there have been 622 requests to GCR.", "From 2006-01-04 15:04:05 to 2006-01-05 15:04:05 [1440 min] there have been 5 requests to GCR.", "From 2006-01-05 15:04:05 to 2006-01-06 15:04:05 [1440 min] there have been 8 requests to GCR."}
>         	            	actual  : []string{"From 2006-01-02 15:04:05 to 2006-01-03 15:04:05 [1440 min] there have been 2 requests to GCR.", "From 2006-01-03 15:04:05 to 2006-01-04 15:14:05 [1440 min] there have been 622 requests to GCR.", "From 2006-01-04 15:14:05 to 2006-01-05 15:14:05 [1440 min] there have been 5 requests to GCR.", "From 2006-01-05 15:14:05 to 2006-01-06 15:14:05 [1440 min] there have been 8 requests to GCR."}
>         	            	
>         	            	Diff:
>         	            	--- Expected
>         	            	+++ Actual
>         	            	@@ -2,5 +2,5 @@
>         	            	  (string) (len=93) "From 2006-01-02 15:04:05 to 2006-01-03 15:04:05 [1440 min] there have been 2 requests to GCR.",
>         	            	- (string) (len=95) "From 2006-01-03 15:04:05 to 2006-01-04 15:04:05 [1440 min] there have been 622 requests to GCR.",
>         	            	- (string) (len=93) "From 2006-01-04 15:04:05 to 2006-01-05 15:04:05 [1440 min] there have been 5 requests to GCR.",
>         	            	- (string) (len=93) "From 2006-01-05 15:04:05 to 2006-01-06 15:04:05 [1440 min] there have been 8 requests to GCR."
>         	            	+ (string) (len=95) "From 2006-01-03 15:04:05 to 2006-01-04 15:14:05 [1440 min] there have been 622 requests to GCR.",
>         	            	+ (string) (len=93) "From 2006-01-04 15:14:05 to 2006-01-05 15:14:05 [1440 min] there have been 5 requests to GCR.",
>         	            	+ (string) (len=93) "From 2006-01-05 15:14:05 to 2006-01-06 15:14:05 [1440 min] there have been 8 requests to GCR."
>         	            	 }
>         	Test:       	TestCycle
>         	Messages:   	The request counter produced malformed logs.
> --- FAIL: TestCycle (0.00s)
> FAIL
> coverage: 100.0% of statements
> FAIL	sigs.k8s.io/k8s-container-image-promoter/legacy/reqcounter	0.015s 
> ```

@tylerferrara ^ FYI

File: legacy/dockerregistry/inventory.go (hunk 1 above)
Comment (justaugustus) [reply legacy/dockerregistry/inventory.go:1194]: @listx -- A lot of this is minimum viable lint fixes, as the amount of cleanup required was more than I was expecting.

I've left a `TODO` atop the function that someone can follow-up on:

```go
// TODO: Can we simplify this to not use switch/case/goto?
```

That said, branchless case statements were the reason for the lint warnings, so I don't think we should move backwards:

```console
legacy/dockerregistry/inventory.go:1191:3: emptyFallthrough: remove empty case containing only fallthrough to default case (gocritic)
		fallthrough
		^
legacy/dockerregistry/inventory.go:1189:3: emptyFallthrough: remove empty case containing only fallthrough to default case (gocritic)
		fallthrough
		^
legacy/dockerregistry/inventory.go:1182:3: emptyFallthrough: replace empty case containing only fallthrough with expression list (gocritic)
		fallthrough
		^
legacy/dockerregistry/inventory.go:1180:3: emptyFallthrough: replace empty case containing only fallthrough with expression list (gocritic)
		fallthrough
		^
legacy/dockerregistry/inventory.go:1178:3: emptyFallthrough: replace empty case containing only fallthrough with expression list (gocritic)
		fallthrough
		^
```

ref: https://github.com/kubernetes-sigs/k8s-container-image-promoter/pull/394#issuecomment-894833433

Comment (listx) [reply legacy/dockerregistry/inventory.go:1194]: If we're going to make the linter happy only to create a TODO comment, in the future I would advise to just turn off the linter locally at that location (and leave a TODO about why we're suppressing the linter). In this case the linter made our code harder to read :(

Comment (justaugustus) [reply legacy/dockerregistry/inventory.go:1194]: The goal of the linting was not to create extraneous TODOs.

> just turn off the linter locally at that location (and leave a TODO about why we're suppressing the linter)

There are several places across this codebase where I do exactly this, but in this case the TODO was left to later assess the entirety of the function (which is why it was placed atop and not next to one of the `case`s).

**Pull request merged**